setter can't handle Callback argument

When we try to have a `Callback` argument in a setter, the runtime panic.
But using a function instead of a setter with the same arguments works correctly.

```diff
diff --git a/src/tests/cbk_test.zig b/src/tests/cbk_test.zig
index 4f7bf0e..c8d18a4 100644
--- a/src/tests/cbk_test.zig
+++ b/src/tests/cbk_test.zig
@@ -59,6 +59,10 @@ pub const Window = struct {
         // ignore the error to let the JS msg
     }
 
+    pub fn get_cbk(_: Window) void {}
+
+    pub fn set_cbk(_: *Window, _: Callback) !void {}
+
     pub fn deinit(_: *Window, _: std.mem.Allocator) void {}
 };
 
@@ -212,4 +216,10 @@ pub fn exec(
         },
     };
     try tests.checkCases(js_env, &cases_cbk_async_with_nat_arg);
+
+    // setter cbk
+    var cases_cbk_setter_arg = [_]tests.Case{
+        .{ .src = "window.cbk =  () => {};", .ex = "undefined" },
+    };
+    try tests.checkCases(js_env, &cases_cbk_setter_arg);
 }
```

```
panic: access of union field 'func_cbk' while field 'prop_cbk' is active                                                                                                                                                         
/home/pierre/wrk/browsercore/vendor/jsruntime-lib/src/engines/v8/generate.zig:320:29: 0xa90b2f in getArgs__anon_22619 (test)                                                                                                     
                    cbk_info.func_cbk,                                                                                                                                                                                           
                            ^                                                                                   
/home/pierre/wrk/browsercore/vendor/jsruntime-lib/src/engines/v8/generate.zig:787:23: 0xa21ad4 in callFunc__anon_21043 (test)
    var args = getArgs(                                                                                         
                      ^                                 
/home/pierre/wrk/browsercore/vendor/jsruntime-lib/src/engines/v8/generate.zig:953:21: 0x9f51f1 in setter (test) 
            callFunc(                                                                                           
                    ^                                                                                           
../../../../v8/src/api/api-arguments-inl.h:332:3: 0x16bf61d in CallAccessorSetter (../../../../v8/src/objects/objects.cc)
../../../../v8/src/objects/objects.cc:1521:34: 0x16bf1ec in SetPropertyWithAccessor (../../../../v8/src/objects/objects.cc)
../../../../v8/src/objects/objects.cc:2560:16: 0x16c6232 in SetPropertyInternal (../../../../v8/src/objects/objects.cc)
../../../../v8/src/objects/objects.cc:2635:9: 0x16c5c27 in SetProperty (../../../../v8/src/objects/objects.cc)  
../../../../v8/src/ic/ic.cc:1851:5: 0x2735fc6 in Store (../../../../v8/src/ic/ic.cc)                            
../../../../v8/src/ic/ic.cc:2802:3: 0x2740741 in __RT_impl_Runtime_StoreIC_Miss (../../../../v8/src/ic/ic.cc)   
../../../../v8/src/ic/ic.cc:2775:1: 0x273fdb7 in Runtime_StoreIC_Miss (../../../../v8/src/ic/ic.cc)             
???:?:?: 0xff65fe in ??? (???)                          
Unwind error at address `:0xff65fe` (error.AddressOutOfRange), trace may be incomplete   
```

--- a/src/engines/v8/callback.zig
+++ b/src/engines/v8/callback.zig
@@ -9,6 +9,7 @@ const NativeContext = internal.NativeContext;
 
 const JSObjectID = @import("v8.zig").JSObjectID;
 const setNativeType = @import("generate.zig").setNativeType;
+const CallbackInfo = @import("generate.zig").CallbackInfo;
 
 // TODO: Make this JS engine agnostic
 // by providing a common interface
@@ -35,7 +36,8 @@ pub const FuncSync = struct {
     pub fn init(
         alloc: std.mem.Allocator,
         comptime func: refl.Func,
-        info: v8.FunctionCallbackInfo,
+        raw_value: ?*const v8.C_Value,
+        info: CallbackInfo,
         isolate: v8.Isolate,
     ) !FuncSync {
 
@@ -54,12 +56,24 @@ pub const FuncSync = struct {
         // var js_args: [func.args_callback_nb]v8.Value = undefined;
         var js_args = try alloc.alloc(v8.Value, func.args_callback_nb);
         for (js_args_indexes, 0..) |index, i| {
-            js_args[i] = info.getArg(@as(u32, @intCast(index - func.index_offset)));
+            js_args[i] = info.getArg(raw_value, index, func.index_offset) orelse unreachable;
         }
 
+        var idx = func.callback_index.?;
+        if (idx > 0) idx = idx - 1; // -1 because of self
+
         // retrieve callback function
-        const js_func_index = func.callback_index.? - func.index_offset - 1; // -1 because of self
-        const js_func_val = info.getArg(js_func_index);
+        const js_func_val = info.getArg(
+            raw_value,
+            idx,
+            func.index_offset,
+        ) orelse unreachable;
+
+        std.debug.print("idx: {d}, offset: {d}, {any}\n", .{
+            func.callback_index.?,
+            idx,
+            js_func_val,
+        });
         if (!js_func_val.isFunction()) {
             return error.JSWrongType;
         }
@@ -113,13 +127,19 @@ pub const Func = struct {
         alloc: std.mem.Allocator,
         nat_ctx: *NativeContext,
         comptime func: refl.Func,
-        info: v8.FunctionCallbackInfo,
+        raw_value: ?*const v8.C_Value,
+        info: CallbackInfo,
         isolate: v8.Isolate,
     ) !Func {
+        var idx = func.callback_index.?;
+        if (idx > 0) idx = idx - 1; // -1 because of self
 
         // retrieve callback function
-        const js_func_index = func.callback_index.? - func.index_offset - 1; // -1 because of self
-        const js_func_val = info.getArg(js_func_index);
+        const js_func_val = info.getArg(
+            raw_value,
+            idx,
+            func.index_offset,
+        ) orelse unreachable;
         if (!js_func_val.isFunction()) {
             return error.JSWrongType;
         }
@@ -148,7 +168,7 @@ pub const Func = struct {
 
             // retrieve callback arguments
             for (js_args_indexes, 0..) |index, i| {
-                const js_arg = info.getArg(@as(u32, @intCast(index - func.index_offset)));
+                const js_arg = info.getArg(raw_value, index, func.index_offset) orelse unreachable;
                 const js_arg_pers = PersistentValue.init(isolate, js_arg);
                 js_args_pers[i] = js_arg_pers;
             }

--- a/src/engines/v8/generate.zig
+++ b/src/engines/v8/generate.zig
@@ -209,7 +209,7 @@ fn getArg(
     return value;
 }
 
-const CallbackInfo = union(enum) {
+pub const CallbackInfo = union(enum) {
     func_cbk: v8.FunctionCallbackInfo,
     prop_cbk: v8.PropertyCallbackInfo,
 
@@ -241,7 +241,7 @@ const CallbackInfo = union(enum) {
         };
     }
 
-    fn getArg(
+    pub fn getArg(
         self: CallbackInfo,
         raw_value: ?*const v8.C_Value,
         index: usize,
@@ -308,22 +308,19 @@ fn getArgs(
 
             // non-variadic arg
             value = switch (arg.T) {
-
-                // special cases for callback arguments
-                // TODO: because thoses functions requires a v8.FunctionCallbackInfo
-                // they will not be available for generators with v8.PropertyCallbackInfo
-                // (getters and setters)
                 cbk.Func => cbk.Func.init(
                     alloc,
                     nat_ctx,
                     func,
-                    cbk_info.func_cbk,
+                    raw_value,
+                    cbk_info,
                     isolate,
                 ) catch unreachable,
                 cbk.FuncSync => cbk.FuncSync.init(
                     alloc,
                     func,
-                    cbk_info.func_cbk,
+                    raw_value,
+                    cbk_info,
                     isolate,
                 ) catch unreachable,
                 cbk.Arg => cbk.Arg{}, // stage1: we need type

--- a/src/tests/cbk_test.zig
+++ b/src/tests/cbk_test.zig
@@ -59,6 +59,12 @@ pub const Window = struct {
         // ignore the error to let the JS msg
     }
 
+    pub fn get_cbk(_: Window) void {}
+
+    pub fn set_cbk(_: *Window, callback: Callback) !void {
+        callback.call(.{}) catch {};
+    }
+
     pub fn deinit(_: *Window, _: std.mem.Allocator) void {}
 };
 
@@ -212,4 +218,12 @@ pub fn exec(
         },
     };
     try tests.checkCases(js_env, &cases_cbk_async_with_nat_arg);
+
+    // setter cbk
+    var cases_cbk_setter_arg = [_]tests.Case{
+        .{ .src = "let v = 0", .ex = "undefined" },
+        .{ .src = "window.cbk =  () => {v++};", .ex = "() => {v++}" },
+        .{ .src = "v", .ex = "1" },
+    };
+    try tests.checkCases(js_env, &cases_cbk_setter_arg);
 }

